### DISCUSSION TITLE
Modbus on ESP32

### DISCUSSION THREAD

**Discussion by nitinuae** [arendst/Tasmota#16979] [Q&A]
Hello
I am using stock esp32 build and assume that modbus support is built in.
Is there a trick to enable modbusbridge?

Here is my output of various tries.

10:59:25.779 CMD: ModbusTCPConnect
10:59:25.790 RSL: RESULT = {"Command":"Unknown"}
10:59:34.133 CMD: ModbusTCPstart
10:59:34.143 RSL: RESULT = {"Command":"Unknown"}
10:59:40.474 CMD: ModbusTCPstart 502
10:59:40.484 RSL: RESULT = {"Command":"Unknown"}
11:00:11.636 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"raw", "count":4}
11:00:11.649 RSL: RESULT = {"ModbusSend":"Done"}
11:00:57.026 CMD: ModbusSerialConfig 3
11:00:57.048 RSL: RESULT = {"ModbusSerialConfig":"8N1"}
11:01:01.401 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"raw", "count":4}
11:01:01.414 RSL: RESULT = {"ModbusSend":"Done"}
11:02:13.664 CMD: ModbusTCPStart
11:02:13.675 RSL: RESULT = {"Command":"Unknown"}

I am trying to use Tasmota as transparent bridge to convert serial RTU to TCP modbus (The other end is connected to rs485 using TTL converter, and correct pins are connected and assigned ModRx and ModTX)

Thank you

**Comment by sfromis** [comment]
While `#define USE_MODBUS_BRIDGE` should indeed be default for the ESP32 base build, that does not mean that `#define USE_MODBUS_BRIDGE_TCP` is included too.

**Reply by nitinuae** [reply]
I read te modbus page throughly - and it states very clearly that both modbus and modbusbridge is included in the default build.
Based on the responses I am getting from Tasmota - either I am doing something wrong or documentation needs to be clear.
I am referring to this page [https://tasmota.github.io/docs/Modbus-Bridge/](url)

**Reply by sfromis** [reply]
The page does clearly indicate that `USE_MODBUS_BRIDGE_TCP` is an extra feature in addition to `USE_MODBUS_BRIDGE` with the latter indeed included in the default base build for ESP32. But yeah, it does not explicitly state that the TCP part is not included, as the main focus of the driver was to provide bridging via MQTT messages.

**Reply by nitinuae** [reply]
Ok - Got it - Compiled with modbustcpbridge.

Now, I am able to start the bridge.
But everytime I try to read any register, tasmota reboots (I get QPC:reset in console before it disappears)

I am able to send the command using the console.
Is there something special to take care of?


Ok - this is what i get in console after reboot
00:00:00.488 BRY: Berry initialized, RAM used=4041 bytes
00:00:00.527 Project tasmota - ASTRA Version 12.2.0(tasmota)-2_0_5(2022-11-02T13:26:49)
00:00:01.308 WIF: Connecting to AP1 Nitin Channel 13 BSSId 5C:92:5E:9C:18:60 in mode 11n as tasmota-C52C70-3184...
00:00:04.328 WIF: Connected
00:00:04.594 HTP: Web server active on tasmota-C52C70-3184 with IP address 192.168.39.190
15:49:10.604 RSL: INFO1 = {"Info1":{"Module":"ASTRA R4G8","Version":"12.2.0(tasmota)","FallbackTopic":"cmnd/DVES_C52C70_fb/","GroupTopic":"cmnd/tasmotas/"}}
15:49:10.619 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-C52C70-3184","IPAddress":"192.168.39.190"}}
15:49:10.626 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":29,"Reason":"StoreProhibited","EPC":"40084de8","EXCVADDR":"00000000","CallChain":["40084de5","40091071","40096735","40162237","40121047","40104495","401552ce","400e6a5f"]},"BootCount":30}}

**Reply by sfromis** [reply]
`QPC: Reset` is no error message, just a normal indication of successful boot. That RestartReason Exception indicates that Tasmota crashed, for unclear reasons. May depend on exact details.....

**Reply by nitinuae** [reply]
Any update on this - I edited my post to have exact log after it booted. Seems like a bug to me.(I havent seen the code - so guessing as it has word storeprohibited)

**Reply by barbudor** [reply]
I think you need to provide more details on what you are doing
- The ModBus device you are using
- All configuration
- The command you are using

And summon the author with kind request for help @jeroenst 😄

**Reply by nitinuae** [reply]
Hello
Modbus device = PDI BCMS (http://www.ausboard.net.au/index_files/BCMS%20Manual-rev4.pdf)
Tasmota complied with modbusbridgetcp (thats the only change). Running on Astra (Robotdyn)

Way to replicate - Start Modbus bridge on port 502.
Use ANY (i used 1 commerical Modscan64, 1 OS Modserver) software to connect. Tasmota console logs that IP x.x.x.x has conncted on port 502. Now I try to read register 1. The moment I do this, Tasmota reboots.

PS - Without starting the modbus tcp bridge, I am able to read all registers etc using modbussend. This proves that there is no compatibilty issue between Tasmota(running on Astra Robotdyn), Cheap RS485-TTL converter and the PDI BCMS from which I am reading registers.

**Reply by jeroenst** [reply]
Before I can do any debugging I need the payload of the tcp packet send (Wireshark trace).

Can you provide that?

**Reply by nitinuae** [reply]
If you babysit me through it. 
I have attached one capture. See if I have done correctly and makes sense.
39.9 is my computer, 39.190 is tasmota
[https://drive.google.com/file/d/1hbXiHN9-WMX2QbCOiGneunIEasoCKUo8/view?usp=share_link](url)

**Reply by jeroenst** [reply]
Thank you for the trace, I will do some investigation

**Reply by jeroenst** [reply]
I think Tasmota fails because an exception is returned from the modbus device instead of normal data.

**Reply by nitinuae** [reply]
I can send regular commands to it using the console.
I am not doing anything special - As I have to discover the registers, I just ask standard windows programs (modbusmaster or modscan64) to read 10 registers (coils).

I can do that successfuly using console. Eg 
ModBusSend {"deviceaddress": 1, "functioncode": 3, "startaddress": 0, "type":"uint16", "count":10}. I get proper responses.

I do this over TCP,
In my programs(eg modbusmaster), start address is 0, size 10 , and I click any function(read coil/read holding/read dicrete etc), tasmota reboots.

**Reply by nitinuae** [reply]
BTW, what is the correct format of reading single coil?
ModBusSend {"deviceaddress": 1, "functioncode": 1, "startaddress": 406, "type":"uint16", "count":1} doesnot return any data. Maybe modbusbridgetcp is creating exception as functioncode 1 isnt returning anything?

From my wireshark, I see that the software is sending function 1. Is functionode = function?

**Reply by jeroenst** [reply]
Yes it is. I will do some testing next week although I don't have an esp32.

**Reply by jeroenst** [reply]
Your modbussend command for a coil is correct you could try type raw or bit instead.

**Reply by nitinuae** [reply]
Same using raw - functioncode 1, no response.
ModBusSend {"deviceaddress": 1, "functioncode": 1, "startaddress": 0, "type":"raw", "count":1}

**Comment by curzon01** [comment]
unfortunatly USE_MODBUS_BRIDGE_TCP excepts on an ESP32:

`ModBusSend {"deviceaddress": 1, "functioncode": 4, "startaddress": 0, "type":"float", "count":1}`

is returning correct value (here phase 1 voltage of a SDM72DM):

```json
{"ModbusReceived":{"DeviceAddress":1,"FunctionCode":4,"StartAddress":0,"Length":7,"Count":1,"Values":[237.2887725830]}}
```

Using `ModbusTCPStart 5000` with USE_MODBUS_BRIDGE_TCP enabled and on client side 

`mbpoll 172.18.10.243 -p 5000 -a 1 -t 4:float -B -0 -1 -r 0x0 -c 1`

returns no valid result (0):

```shell
Protocol configuration: Modbus TCP
Slave configuration...: address = [1]
                        start reference = 0, count = 1
Communication.........: 172.18.10.243, port 5000, t/o 1.00 s, poll rate 1000 ms
Data type.............: 32-bit float (big endian), output (holding) register table

-- Polling slave 1...
[0]: 	0
```

because ESP32 crashs with exception

```
Stack smashing protect failure!

abort() was called at PC 0x40163ea3 on core 1

Backtrace: 0x40084941:0x3ffb25c0 0x40090c35:0x3ffb25e0 0x40096411:0x3ffb2600 0x40163ea3:0x3ffb2680 0x40122dbb:0x3ffb26a0 0x40106221:0x3ffb2730 0x401578d2:0x3ffb2750 0x400e746f:0x3ffb2820
```

exception trace says:
```shell
0x40163dbb: __stack_chk_fail at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/stack_check.c:36
0x40084941: panic_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/panic.c:417
0x40090c35: esp_system_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/esp_system.c:119
0x40096411: abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/newlib/abort.c:46
0x40163dbb: __stack_chk_fail at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/stack_check.c:36
0x40122d87: Xdrv63(unsigned char) at ??:?
0x40106221: XdrvCall(unsigned char) at ??:?
0x401577ea: loop() at ??:?
0x400e746f: loopTask(void*) at :?
```

restart info was
```json
{"Info3":{"RestartReason":{"Exception":29,"Reason":"StoreProhibited","EPC":"40084944","EXCVADDR":"00000000","CallChain":["40084941","40090c35","40096411","40163dbb","40122d87","40106221","401577ea","400e746f"]},"BootCount":72}}
```

**Reply by jeroenst** [reply]
I'm curious if this only happens on esp32.

On esp8266 it works fine afaik.

**Reply by curzon01** [reply]
Yes, this only happens on an ESP32, it works on an ESP8266

**Reply by jeroenst** [reply]
Ah, ok, I didn't develop it for an esp32, because the building failed anyway when I coded it. Later something has changed by somebody else which made it build on esp32 also.

I don't have an esp32 laying around to debug this, can you use an esp866 instead for the moment?

**Reply by curzon01** [reply]
yes, no problem. I'll see if I can find the issue...

**Reply by jeroenst** [reply]
That would be awesome!

**Reply by curzon01** [reply]
it was simply a buffer overflow due to a too small array, see PR #17007

**Reply by jeroenst** [reply]
Good job! Thank you for helping!
